File: src/main/kotlin/com/cfm/storage/StorageService.kt
package com.cfm.storage

import com.cfm.model.FileModel
import kotlinx.coroutines.future.await
import software.amazon.awssdk.core.async.AsyncRequestBody
import software.amazon.awssdk.core.async.AsyncResponseTransformer
import software.amazon.awssdk.services.s3.S3AsyncClient
import software.amazon.awssdk.services.s3.model.*
import software.amazon.awssdk.services.s3.presigner.model.GetObjectPresignRequest
import java.io.InputStream
import java.nio.ByteBuffer
import java.time.Duration
import java.util.*

/**
 * Service for handling file storage operations using AWS S3
 */
class StorageService(private val s3Client: S3AsyncClient, private val bucketName: String) {

    /**
     * Uploads a file to S3 storage
     *
     * @param fileName The name of the file
     * @param contentType The content type (MIME type) of the file
     * @param inputStream The file content as an input stream
     * @param metadata Additional metadata to store with the file
     * @return The S3 object key and version ID
     */
    suspend fun uploadFile(
        fileName: String,
        contentType: String,
        inputStream: InputStream,
        metadata: Map<String, String> = emptyMap()
    ): Pair<String, String> {
        val objectKey = generateObjectKey(fileName)

        val putObjectRequest = PutObjectRequest.builder()
            .bucket(bucketName)
            .key(objectKey)
            .contentType(contentType)
            .metadata(metadata)
            .build()

        val fileBytes = inputStream.readAllBytes()
        val response = s3Client.putObject(
            putObjectRequest,
            AsyncRequestBody.fromBytes(fileBytes)
        ).await()

        return Pair(objectKey, response.versionId())
    }

    /**
     * Downloads a file from S3 storage
     *
     * @param objectKey The S3 object key
     * @param versionId Optional version ID (if null, the latest version is downloaded)
     * @return The file content as a ByteBuffer
     */
    suspend fun downloadFile(objectKey: String, versionId: String? = null): ByteBuffer {
        val getObjectRequest = GetObjectRequest.builder()
            .bucket(bucketName)
            .key(objectKey)
            .apply {
                if (versionId != null) versionId(versionId)
            }
            .build()

        return s3Client.getObject(
            getObjectRequest,
            AsyncResponseTransformer.toBytes()
        ).await().asByteBuffer()
    }

    /**
     * Generates a presigned URL for temporary file access
     *
     * @param objectKey The S3 object key
     * @param durationSeconds The duration in seconds for which the URL is valid
     * @param versionId Optional version ID
     * @return The presigned URL as a string
     */
    fun generatePresignedUrl(
        objectKey: String,
        durationSeconds: Long = 3600,
        versionId: String? = null
    ): String {
        val getObjectRequest = GetObjectRequest.builder()
            .bucket(bucketName)
            .key(objectKey)
            .apply {
                if (versionId != null) versionId(versionId)
            }
            .build()

        val presignRequest = GetObjectPresignRequest.builder()
            .signatureDuration(Duration.ofSeconds(durationSeconds))
            .getObjectRequest(getObjectRequest)
            .build()

        return s3Client.presignGetObject(presignRequest).url().toString()
    }

    /**
     * Deletes a file from S3 storage
     *
     * @param objectKey The S3 object key
     * @return True if deletion was successful
     */
    suspend fun deleteFile(objectKey: String): Boolean {
        val deleteObjectRequest = DeleteObjectRequest.builder()
            .bucket(bucketName)
            .key(objectKey)
            .build()

        val response = s3Client.deleteObject(deleteObjectRequest).await()
        return response.deleteMarker()
    }

    /**
     * Lists all file versions for the given object key
     *
     * @param objectKey The S3 object key
     * @return List of version IDs and their creation timestamps
     */
    suspend fun listFileVersions(objectKey: String): List<Pair<String, String>> {
        val listVersionsRequest = ListObjectVersionsRequest.builder()
            .bucket(bucketName)
            .prefix(objectKey)
            .build()

        val response = s3Client.listObjectVersions(listVersionsRequest).await()

        return response.versions().map { version ->
            Pair(version.versionId(), version.lastModified().toString())
        }
    }

    /**
     * Generates a unique object key for S3 storage based on the file name
     *
     * @param fileName The original file name
     * @return A unique S3 object key
     */
    private fun generateObjectKey(fileName: String): String {
        val uuid = UUID.randomUUID().toString()
        val sanitizedFileName = fileName.replace("[^a-zA-Z0-9.-]".toRegex(), "_")

        // Create a path with year/month structure for better organization
        val today = Calendar.getInstance()
        val year = today.get(Calendar.YEAR)
        val month = today.get(Calendar.MONTH) + 1

        return "$year/$month/$uuid-$sanitizedFileName"
    }
}

File: cloud_file_manager/storage/storage_service.py
import io
import logging
import uuid
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Tuple, BinaryIO

import boto3
from botocore.exceptions import ClientError

logger = logging.getLogger(__name__)


class StorageService:
    """Service for handling file storage operations using AWS S3"""

    def __init__(self, bucket_name: str, region_name: str = "us-east-1"):
        self.bucket_name = bucket_name
        self.region_name = region_name
        self.s3_client = boto3.client("s3", region_name=region_name)
        self.s3_resource = boto3.resource("s3", region_name=region_name)

    def upload_file(
            self,
            file_name: str,
            content_type: str,
            file_data: BinaryIO,
            metadata: Dict[str, str] = None
    ) -> Tuple[str, str]:
        """
        Uploads a file to S3 storage

        Args:
            file_name: The name of the file
            content_type: The content type (MIME type) of the file
            file_data: The file content as a file-like object
            metadata: Additional metadata to store with the file

        Returns:
            Tuple containing the S3 object key and version ID
        """
        if metadata is None:
            metadata = {}

        # Generate a unique object key
        object_key = self._generate_object_key(file_name)

        try:
            # Upload the file to S3
            response = self.s3_client.upload_fileobj(
                file_data,
                self.bucket_name,
                object_key,
                ExtraArgs={
                    "ContentType": content_type,
                    "Metadata": metadata
                }
            )

            # Get the version ID
            head_response = self.s3_client.head_object(
                Bucket=self.bucket_name,
                Key=object_key
            )
            version_id = head_response.get("VersionId", "")

            return object_key, version_id

        except ClientError as e:
            logger.error(f"Error uploading file to S3: {e}")
            raise

    def download_file(self, object_key: str, version_id: Optional[str] = None) -> bytes:
        """
        Downloads a file from S3 storage

        Args:
            object_key: The S3 object key
            version_id: Optional version ID (if None, the latest version is downloaded)

        Returns:
            The file content as bytes
        """
        try:
            # Create a BytesIO object to store the downloaded file
            file_data = io.BytesIO()

            # Download the file from S3
            kwargs = {
                "Bucket": self.bucket_name,
                "Key": object_key
            }

            if version_id:
                kwargs["VersionId"] = version_id

            self.s3_client.download_fileobj(
                **kwargs,
                Fileobj=file_data
            )

            # Reset the file pointer to the beginning and return the bytes
            file_data.seek(0)
            return file_data.read()

        except ClientError as e:
            logger.error(f"Error downloading file from S3: {e}")
            raise

    def generate_presigned_url(
            self,
            object_key: str,
            duration_seconds: int = 3600,
            version_id: Optional[str] = None
    ) -> str:
        """
        Generates a presigned URL for temporary file access

        Args:
            object_key: The S3 object key
            duration_seconds: The duration in seconds for which the URL is valid
            version_id: Optional version ID

        Returns:
            The presigned URL as a string
        """
        try:
            # Generate a presigned URL for the object
            params = {
                "Bucket": self.bucket_name,
                "Key": object_key
            }

            if version_id:
                params["VersionId"] = version_id

            url = self.s3_client.generate_presigned_url(
                ClientMethod="get_object",
                Params=params,
                ExpiresIn=duration_seconds
            )

            return url

        except ClientError as e:
            logger.error(f"Error generating presigned URL: {e}")
            raise

    def delete_file(self, object_key: str) -> bool:
        """
        Deletes a file from S3 storage

        Args:
            object_key: The S3 object key

        Returns:
            True if deletion was successful
        """
        try:
            # Delete the object from S3
            response = self.s3_client.delete_object(
                Bucket=self.bucket_name,
                Key=object_key
            )

            # Check if deletion was successful
            return response.get("DeleteMarker", False)

        except ClientError as e:
            logger.error(f"Error deleting file from S3: {e}")
            raise

    def list_file_versions(self, object_key: str) -> List[Tuple[str, str]]:
        """
        Lists all file versions for the given object key

        Args:
            object_key: The S3 object key

        Returns:
            List of tuples containing version IDs and their creation timestamps
        """
        try:
            # List versions of the object
            response = self.s3_client.list_object_versions(
                Bucket=self.bucket_name,
                Prefix=object_key
            )

            # Extract version information
            versions = []
            for version in response.get("Versions", []):
                if version.get("Key") == object_key:
                    versions.append((
                        version.get("VersionId", ""),
                        version.get("LastModified", datetime.now()).isoformat()
                    ))

            return versions

        except ClientError as e:
            logger.error(f"Error listing file versions: {e}")
            raise

    def _generate_object_key(self, file_name: str) -> str:
        """
        Generates a unique object key for S3 storage based on the file name

        Args:
            file_name: The original file name

        Returns:
            A unique S3 object key
        """
        # Generate a UUID
        unique_id = str(uuid.uuid4())

        # Sanitize the file name (remove special characters)
        sanitized_name = "".join(c if c.isalnum() or c in ".-_" else "_" for c in file_name)

        # Create a path with year/month structure for better organization
        now = datetime.now()
        year = now.year
        month = now.month

        return f"{year}/{month}/{unique_id}-{sanitized_name}"